File: crm/lib/agent/tour-follow-up.ts
/**
 * Follow-up SMS after a showing actually finishes.
 *
 * The tour PATCH already fires `tour_completed`. This writes one short
 * text in the assigned realtor's voice and sends it through Telnyx. The
 * follow-up is an ask — how did it feel, do they want to talk next —
 * never a claim that a deal closed or a time is booked / reserved /
 * locked / held.
 *
 * No approval inbox. No pending persist. Missing credentials fail.
 */

import { supabase } from '@/lib/supabase';
import { logger } from '@/lib/logger';
import {
  firstNameOf,
  listingFromTourOrContact,
  normalizeTone,
  sendAutonomousSms,
  type AgentTone,
  type FirstTouchVoice,
} from '@/lib/agent/first-touch';
import { isTourCompletedEvent } from '@/lib/agent/trigger-policy';

export type TourFollowUpAction = 'sent' | 'deduped' | 'filled' | 'skipped';

export interface TourFollowUpDraftResult {
  action: TourFollowUpAction;
  draftId?: string;
  contactId: string;
  channel: 'sms';
  status: 'sent' | 'skipped';
  content: string;
  sent: boolean;
  reason?: string;
}

export interface ComposeTourFollowUpInput {
  contactFirstName: string;
  voice: FirstTouchVoice;
  property?: string;
}

export interface DraftTourFollowUpInput {
  spaceId: string;
  contactId: string;
  tourId?: string;
  now?: Date;
}

const DEDUPE_WINDOW_HOURS = 48;
const REASON_MARK = 'Tour-completed follow-up SMS';
const REASON = 'Tour-completed follow-up SMS — ask how the showing felt. Sent.';

export function isTourCompletedFollowUpEvent(event: string): boolean {
  return isTourCompletedEvent(event);
}

export function isTourFollowUpDraft(row: { reasoning?: string | null }): boolean {
  return (row.reasoning ?? '').includes(REASON_MARK);
}

export function assertSentDraftPersist(row: { status: string }): void {
  if (row.status === 'pending') {
    throw new Error('tour-follow-up persist must not stay pending');
  }
  if (row.status !== 'sent') {
    throw new Error('tour-follow-up persist must be sent');
  }
}

export function composeTourFollowUpSms(input: ComposeTourFollowUpInput): string {
  const lead = firstNameOf(input.contactFirstName, 'there');
  const agent = firstNameOf(input.voice.agentFirstName, '');
  const who = agent || input.voice.businessName?.trim() || 'I';
  const property = input.property?.trim();
  const place = property || 'the showing';

  let text: string;
  switch (input.voice.tone) {
    case 'direct':
      text = `Hi ${lead} — ${who} here. Thoughts on ${place}? Ready to talk next?`;
      break;
    case 'formal':
      text = `Hello ${lead}, this is ${who}. How was ${
        property ? `the showing at ${property}` : 'the showing'
      }? Would you like to discuss next steps?`;
      break;
    case 'casual':
      text = `Hey ${lead}! ${who} here — how'd ${place} feel? Want to chat next steps?`;
      break;
    case 'warm':
    default:
      text = `Hey ${lead}, this is ${who}. How did ${place} feel? Want to talk next steps?`;
      break;
  }

  const content = text.replace(/\s+/g, ' ').trim();
  assertValidTourFollowUpText(content, { agentToken: who, tone: input.voice.tone });
  return content;
}

export function assertValidTourFollowUpText(
  content: string,
  opts: { agentToken: string; tone: AgentTone },
): void {
  if (!content.trim()) {
    throw new Error('tour-follow-up draft is empty');
  }
  if (/\b(sent|delivered|auto-?sent)\b/i.test(content)) {
    throw new Error('tour-follow-up draft claims it was sent');
  }
  if (
    /\b(booked|live|reserved|locked|held)\b/i.test(content) ||
    /\bis held\b/i.test(content) ||
    /\bi['’]ll lock\b/i.test(content) ||
    /\bsee you (then|there)\b/i.test(content)
  ) {
    throw new Error('tour-follow-up draft claims the showing is booked');
  }
  if (
    /\b(deal is closed|we closed|closed the deal|it['’]s closed|you(?:'ve| have)? closed)\b/i.test(
      content,
    )
  ) {
    throw new Error('tour-follow-up draft claims the deal is closed');
  }
  if (/\bchippy\b/i.test(content)) {
    throw new Error('tour-follow-up draft used the wrong brand spelling');
  }
  if (opts.agentToken && !content.includes(opts.agentToken)) {
    throw new Error('tour-follow-up draft is not in the assigned agent voice');
  }
  const toneMarks: Record<AgentTone, RegExp> = {
    warm: /this is|want to talk next/i,
    direct: /thoughts on|ready to talk next/i,
    formal: /would you like to discuss/i,
    casual: /how'd|want to chat next/i,
  };
  if (!toneMarks[opts.tone].test(content)) {
    throw new Error(`tour-follow-up draft is not in the ${opts.tone} voice`);
  }
}

function skipped(contactId: string, reason: string): TourFollowUpDraftResult {
  return {
    action: 'skipped',
    contactId,
    channel: 'sms',
    status: 'skipped',
    content: '',
    sent: false,
    reason,
  };
}

type DraftRow = {
  id: string;
  content: string | null;
  status: string;
  channel: string;
  reasoning?: string | null;
  createdAt?: string;
};

export async function draftTourFollowUpForContact(
  input: DraftTourFollowUpInput,
): Promise<TourFollowUpDraftResult> {
  if (!input.spaceId || !input.contactId) {
    throw new Error('spaceId and contactId are required');
  }

  const now = input.now ?? new Date();
  const { data: contact, error: contactError } = await supabase
    .from('Contact')
    .select('id,name,phone,address,preferences,properties,applicationData,spaceId')
    .eq('id', input.contactId)
    .eq('spaceId', input.spaceId)
    .maybeSingle();

  if (contactError) {
    logger.error('[tour-follow-up] contact lookup failed', { spaceId: input.spaceId }, contactError);
    throw new Error('Contact lookup failed');
  }
  if (!contact) {
    throw new Error('Contact not found in space');
  }

  let tourQuery = supabase
    .from('Tour')
    .select('id,status,propertyAddress,contactId,spaceId,updatedAt')
    .eq('spaceId', input.spaceId)
    .eq('status', 'completed');
  tourQuery = input.tourId
    ? tourQuery.eq('id', input.tourId)
    : tourQuery.eq('contactId', input.contactId);
  const { data: tourRows } = await tourQuery.order('updatedAt', { ascending: false }).limit(1);
  const tour = ((tourRows ?? [])[0] ?? null) as {
    id: string;
    status: string;
    propertyAddress?: string | null;
    contactId?: string | null;
    spaceId?: string;
  } | null;

  if (!tour || tour.status !== 'completed') {
    return skipped(input.contactId, 'no_completed_tour');
  }
  if (tour.contactId && tour.contactId !== input.contactId) {
    return skipped(input.contactId, 'tour_contact_mismatch');
  }

  const cutoff = new Date(now.getTime() - DEDUPE_WINDOW_HOURS * 60 * 60 * 1000).toISOString();
  const { data: existingRows } = await supabase
    .from('AgentDraft')
    .select('id,content,status,channel,reasoning,createdAt')
    .eq('spaceId', input.spaceId)
    .eq('contactId', input.contactId)
    .eq('channel', 'sms')
    .gte('createdAt', cutoff)
    .order('createdAt', { ascending: false })
    .limit(20);

  const drafts = (existingRows ?? []) as DraftRow[];
  const alreadySent = drafts.find(
    (row) => isTourFollowUpDraft(row) && row.status === 'sent' && Boolean(row.content?.trim()),
  );
  const emptyStub = drafts.find((row) => isTourFollowUpDraft(row) && !row.content?.trim());

  const [profileRes, settingRes, spaceRes] = await Promise.all([
    supabase
      .from('AIUserProfile')
      .select('displayName,communicationTone')
      .eq('spaceId', input.spaceId)
      .maybeSingle(),
    supabase.from('SpaceSetting').select('businessName').eq('spaceId', input.spaceId).maybeSingle(),
    supabase.from('Space').select('name').eq('id', input.spaceId).maybeSingle(),
  ]);

  const profile = profileRes.data as { displayName?: string | null; communicationTone?: string | null } | null;
  const setting = settingRes.data as { businessName?: string | null } | null;
  const space = spaceRes.data as { name?: string | null } | null;

  const voice: FirstTouchVoice = {
    tone: normalizeTone(profile?.communicationTone),
    agentFirstName: firstNameOf(profile?.displayName, ''),
    businessName: setting?.businessName?.trim() || space?.name?.trim() || undefined,
  };
  if (!voice.agentFirstName && voice.businessName) {
    voice.agentFirstName = voice.businessName;
  }
  if (!voice.agentFirstName) {
    voice.agentFirstName = 'I';
  }

  const content = composeTourFollowUpSms({
    contactFirstName: firstNameOf(contact.name, 'there'),
    voice,
    property: listingFromTourOrContact(tour, contact),
  });
  if (!content.trim()) {
    throw new Error('tour-follow-up draft is empty');
  }

  if (alreadySent) {
    return {
      action: 'deduped',
      draftId: alreadySent.id,
      contactId: input.contactId,
      channel: 'sms',
      status: 'sent',
      content: alreadySent.content!,
      sent: true,
    };
  }

  await sendAutonomousSms({
    to: (contact as { phone?: string | null }).phone,
    body: content,
    label: 'tour-follow-up',
  });

  const expiresAt = new Date(now.getTime() + 7 * 24 * 60 * 60 * 1000).toISOString();
  const row = {
    id: emptyStub?.id ?? crypto.randomUUID(),
    spaceId: input.spaceId,
    contactId: input.contactId,
    channel: 'sms' as const,
    content,
    reasoning: REASON,
    priority: 82,
    status: 'sent' as const,
    expiresAt,
    updatedAt: now.toISOString(),
  };
  assertSentDraftPersist(row);

  if (emptyStub) {
    const update = {
      content: row.content,
      reasoning: row.reasoning,
      priority: row.priority,
      status: 'sent' as const,
      expiresAt: row.expiresAt,
      updatedAt: row.updatedAt,
    };
    assertSentDraftPersist(update);
    const { error: updateError } = await supabase
      .from('AgentDraft')
      .update(update)
      .eq('id', emptyStub.id)
      .eq('spaceId', input.spaceId);
    if (updateError) {
      logger.error('[tour-follow-up] failed to record sent SMS', { spaceId: input.spaceId }, updateError);
      throw new Error('Failed to record tour-follow-up SMS');
    }
    return {
      action: 'filled',
      draftId: emptyStub.id,
      contactId: input.contactId,
      channel: 'sms',
      status: 'sent',
      content,
      sent: true,
    };
  }

  const { data: inserted, error: insertError } = await supabase
    .from('AgentDraft')
    .insert(row)
    .select('id')
    .maybeSingle();
  if (insertError) {
    logger.error('[tour-follow-up] insert failed', { spaceId: input.spaceId }, insertError);
    throw new Error('Failed to record tour-follow-up SMS');
  }

  return {
    action: 'sent',
    draftId: (inserted as { id?: string } | null)?.id ?? row.id,
    contactId: input.contactId,
    channel: 'sms',
    status: 'sent',
    content,
    sent: true,
  };
}

File: crm/agent/tour_follow_up.py
"""Follow-up SMS after a showing finishes — autonomous-run backstop.

The TypeScript event trigger (`lib/agent/tour-follow-up.ts`) sends as
soon as the tour is marked completed. This module does the same job when
the autonomous run drains the trigger queue.

Sends through Telnyx. Status is sent. The text is an ask — how did it
feel, do they want to talk next — never a claim that a deal closed or a
time is booked / reserved / locked / held. Missing credentials fail —
they do not fall back to a pending draft.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timedelta, timezone
from typing import Any

from first_touch import first_name_of, normalize_tone, send_sms

TOUR_COMPLETED_EVENT = "tour_completed"
_DEDUPE_WINDOW_HOURS = 48
_REASON_MARK = "Tour-completed follow-up SMS"
_REASON = "Tour-completed follow-up SMS — ask how the showing felt. Sent."

_TONE_MARKS = {
    "warm": ("this is", "want to talk next"),
    "direct": ("thoughts on", "ready to talk next"),
    "formal": ("would you like to discuss",),
    "casual": ("how'd", "want to chat next"),
}


def is_tour_completed_event(event: str | None) -> bool:
    return event == TOUR_COMPLETED_EVENT


def is_tour_follow_up_draft(row: dict[str, Any] | None) -> bool:
    return bool(row) and _REASON_MARK in (row.get("reasoning") or "")


def assert_sent_draft_persist(row: dict[str, Any]) -> None:
    if row.get("status") == "pending":
        raise ValueError("tour-follow-up persist must not stay pending")
    if row.get("status") != "sent":
        raise ValueError("tour-follow-up persist must be sent")


def compose_tour_follow_up_sms(
    *,
    contact_first_name: str,
    agent_first_name: str,
    tone: str,
    property_name: str | None = None,
    business_name: str | None = None,
) -> str:
    lead = first_name_of(contact_first_name, "there")
    who = first_name_of(agent_first_name, "") or (business_name or "").strip() or "I"
    property_label = (property_name or "").strip()
    place = property_label or "the showing"
    tone_key = normalize_tone(tone)

    if tone_key == "direct":
        text = f"Hi {lead} — {who} here. Thoughts on {place}? Ready to talk next?"
    elif tone_key == "formal":
        shown = f"the showing at {property_label}" if property_label else "the showing"
        text = f"Hello {lead}, this is {who}. How was {shown}? Would you like to discuss next steps?"
    elif tone_key == "casual":
        text = f"Hey {lead}! {who} here — how'd {place} feel? Want to chat next steps?"
    else:
        text = f"Hey {lead}, this is {who}. How did {place} feel? Want to talk next steps?"

    content = " ".join(text.split()).strip()
    assert_valid_tour_follow_up_text(content, agent_token=who, tone=tone_key)
    return content


def assert_valid_tour_follow_up_text(
    content: str,
    *,
    agent_token: str,
    tone: str,
) -> None:
    if not content.strip():
        raise ValueError("tour-follow-up draft is empty")
    lower = content.lower()
    if any(word in lower for word in ("sent", "delivered", "auto-sent", "autosent")):
        raise ValueError("tour-follow-up draft claims it was sent")
    if any(
        word in lower
        for word in (
            "booked",
            "live",
            "reserved",
            "locked",
            "held",
            "is held",
            "i'll lock",
            "i’ll lock",
            "see you then",
            "see you there",
        )
    ):
        raise ValueError("tour-follow-up draft claims the showing is booked")
    if any(
        phrase in lower
        for phrase in (
            "deal is closed",
            "we closed",
            "closed the deal",
            "it's closed",
            "it’s closed",
            "you've closed",
            "you have closed",
        )
    ):
        raise ValueError("tour-follow-up draft claims the deal is closed")
    if "chippy" in lower:
        raise ValueError("tour-follow-up draft used the wrong brand spelling")
    if agent_token and agent_token not in content:
        raise ValueError("tour-follow-up draft is not in the assigned agent voice")
    marks = _TONE_MARKS[normalize_tone(tone)]
    if not any(mark in lower for mark in marks):
        raise ValueError(f"tour-follow-up draft is not in the {tone} voice")


def tour_follow_up_instruction(triggers: list[dict]) -> str | None:
    """Extra opening-prompt block for tour-completed triggers."""
    completed = [
        t
        for t in triggers
        if is_tour_completed_event(t.get("event")) and t.get("contactId")
    ]
    if not completed:
        return None
    lines = []
    for t in completed:
        line = f"- tour_completed contactId: {t['contactId']}"
        if t.get("tourId"):
            line += f" tourId: {t['tourId']}"
        lines.append(line)
    return (
        "TOUR FOLLOW-UP — a showing just finished. Send one short SMS "
        "in the assigned realtor's voice asking how it felt and whether "
        "they want to talk next steps. Do not park a draft. Do not wait "
        "for approval. Do not claim a deal is closed or a time is booked, "
        "reserved, locked, or held.\n" + "\n".join(lines)
    )


def _skipped(contact_id: str, reason: str) -> dict[str, Any]:
    return {
        "action": "skipped",
        "contactId": contact_id,
        "status": "skipped",
        "channel": "sms",
        "content": "",
        "sent": False,
        "reason": reason,
    }


def _property_from_tour_or_contact(
    tour: dict[str, Any] | None,
    contact: dict[str, Any],
) -> str | None:
    if tour:
        address = (tour.get("propertyAddress") or "").strip()
        if address:
            return address
    address = (contact.get("address") or "").strip()
    if address:
        return address
    for item in contact.get("properties") or []:
        if isinstance(item, str) and item.strip():
            return item.strip()
    data = contact.get("applicationData")
    if isinstance(data, dict):
        for key in (
            "address",
            "propertyAddress",
            "listingAddress",
            "property",
            "interestedProperty",
        ):
            value = data.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()
    return None


async def ensure_tour_follow_up_draft(
    space_id: str,
    contact_id: str,
    *,
    tour_id: str | None = None,
    now: datetime | None = None,
) -> dict[str, Any]:
    """Send a tour-follow-up SMS. Persists sent. Never parks pending."""
    from db import supabase
    from tools.base import with_retry

    when = now or datetime.now(timezone.utc)
    db = await supabase()

    check = await (
        db.table("Contact")
        .select("id,name,phone,address,properties,applicationData")
        .eq("id", contact_id)
        .eq("spaceId", space_id)
        .maybe_single()
        .execute()
    )
    if not check.data:
        return {"error": "Contact not found in space", "sent": False}

    contact = check.data
    tour_q = (
        db.table("Tour")
        .select("id,status,propertyAddress,contactId,spaceId,updatedAt")
        .eq("spaceId", space_id)
        .eq("status", "completed")
    )
    if tour_id:
        tour_q = tour_q.eq("id", tour_id)
    else:
        tour_q = tour_q.eq("contactId", contact_id)
    tours = await tour_q.order("updatedAt", desc=True).limit(1).execute()
    tour = tours.data[0] if tours.data else None
    if not tour or tour.get("status") != "completed":
        return _skipped(contact_id, "no_completed_tour")
    if tour.get("contactId") and tour.get("contactId") != contact_id:
        return _skipped(contact_id, "tour_contact_mismatch")

    cutoff = (when - timedelta(hours=_DEDUPE_WINDOW_HOURS)).isoformat()
    existing = await (
        db.table("AgentDraft")
        .select("id,content,status,channel,reasoning,createdAt")
        .eq("spaceId", space_id)
        .eq("contactId", contact_id)
        .eq("channel", "sms")
        .gte("createdAt", cutoff)
        .order("createdAt", desc=True)
        .limit(20)
        .execute()
    )
    drafts = existing.data or []
    already_sent = next(
        (
            row
            for row in drafts
            if is_tour_follow_up_draft(row)
            and row.get("status") == "sent"
            and (row.get("content") or "").strip()
        ),
        None,
    )
    empty_stub = next(
        (
            row
            for row in drafts
            if is_tour_follow_up_draft(row) and not (row.get("content") or "").strip()
        ),
        None,
    )

    profile_res = await (
        db.table("AIUserProfile")
        .select("displayName,communicationTone")
        .eq("spaceId", space_id)
        .maybe_single()
        .execute()
    )
    setting_res = await (
        db.table("SpaceSetting")
        .select("businessName")
        .eq("spaceId", space_id)
        .maybe_single()
        .execute()
    )
    space_res = await (
        db.table("Space").select("name").eq("id", space_id).maybe_single().execute()
    )
    profile = profile_res.data or {}
    setting = setting_res.data or {}
    space = space_res.data or {}

    agent_name = first_name_of(profile.get("displayName"), "")
    business = (setting.get("businessName") or space.get("name") or "").strip() or None
    if not agent_name:
        agent_name = business or "I"

    content = compose_tour_follow_up_sms(
        contact_first_name=first_name_of(contact.get("name"), "there"),
        agent_first_name=agent_name,
        tone=normalize_tone(profile.get("communicationTone")),
        property_name=_property_from_tour_or_contact(tour, contact),
        business_name=business,
    )
    if not content.strip():
        raise ValueError("tour-follow-up draft is empty")

    if already_sent:
        return {
            "action": "deduped",
            "draftId": already_sent["id"],
            "status": "sent",
            "channel": "sms",
            "content": already_sent["content"],
            "sent": True,
        }

    await send_sms(to=contact.get("phone"), body=content, label="tour-follow-up")

    expires_at = (when + timedelta(days=7)).isoformat()
    if empty_stub:
        update = {
            "content": content,
            "reasoning": _REASON,
            "priority": 82,
            "status": "sent",
            "expiresAt": expires_at,
            "updatedAt": when.isoformat(),
        }
        assert_sent_draft_persist(update)
        await with_retry(
            lambda: db.table("AgentDraft")
            .update(update)
            .eq("id", empty_stub["id"])
            .eq("spaceId", space_id)
            .execute()
        )
        return {
            "action": "filled",
            "draftId": empty_stub["id"],
            "status": "sent",
            "channel": "sms",
            "content": content,
            "sent": True,
        }

    draft = {
        "id": str(uuid.uuid4()),
        "spaceId": space_id,
        "contactId": contact_id,
        "channel": "sms",
        "content": content,
        "reasoning": _REASON,
        "priority": 82,
        "status": "sent",
        "expiresAt": expires_at,
    }
    assert_sent_draft_persist(draft)
    result = await with_retry(lambda: db.table("AgentDraft").insert(draft).execute())
    created = result.data[0] if result.data else draft
    return {
        "action": "sent",
        "draftId": created.get("id", draft["id"]),
        "status": "sent",
        "channel": "sms",
        "content": content,
        "sent": True,
    }
